### ISSUE TITLE
 Requesting Advice on NF Methods

### ISSUE THREAD

**Issue by kayuksel** [VLL-HD/FrEIA#77]
I am working on a project where I sample a set of n-dimensional points from a Gaussian distribution (of learnt parameters) as follows and then evaluate those points based on a loss function to update model parameters with gradient descent.

```
mu, std = self.lin_1(z), self.lin_2(z)
eps = torch.Tensor(*img_shape).normal_()
return self.act((eps.cuda() * std) + mu)
```

I would like to transform the Gaussian distribution for being able to sample those points from a more complex learnt distribution. In other words, the model needs to learn how to best transform points obtained from the Gaussian distribution.

I would be glad if you can suggest the best normalizing flows method (transform) to employ considering the following scalability requirements (whether or not it is available in this repo). Thank you very much in advance for your suggestion.

- I am sampling 100K-dimensional points with a batch-size of 5K; hence, the scalability is crucial.
- The method should be memory efficient and fast to train on a RTX series desktop Nvidia GPU.
- There should not ideally be an additional regularization parameter to my current loss function.
- Expressiveness of the method is not as important as scalability and robustness in the training.

**Comment by MattSkiff** [comment]
I'm afraid I can't recommend a specific architecture based on your requirements - that would require an extensive literature survey. I can provide some advice around feasibility and notion of using a custom loss function. 

I've been using this repository for a little while now. The dimensionality of the flow I am training was 800x600x1, so 480k -> therefore, it shouldn't be overly difficult to learn a problem of dim 100k. However, I found that for reasonably deep flows and large batch sizes, I had use A100s, which are quite serious GPUs. Your memory requirements are going to come down to how deep you want the flow to be and the size of the subnetworks that you choose.

While you can train NFs with other loss functions, typically they are trained with the NLL loss function. This is for theoretical reasons in terms of defining the sequence of invertible, differentiable mappings (or transforms) between the base measure and the data distribution - use of the NLL corresponds to minimizing the KL divergence between the base and data distributions. 

The use of this loss is also often considered a key advantage, as it removes the need for complex, custom loss functions (common in computer vision) and training has been observed to be quick and stable (which I also found in my case, once I used a sufficiently large batch size).  With that said, there is existing published research using other loss functions. 'Dual Glow' is one example (modality transfer).

Hope this helps,

Cheers,
Matthew
